sys/usb/usbus/cdc/ecm/cdc_ecm.c

```diff
@@ -0,0 +1,386 @@
+/*
+ * Copyright (C) 2019 Koen Zandberg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup usbus_cdc_eem
+ * @{
+ * @file USBUS implementation for ethernet emulation model
+ *
+ * @author  Koen Zandberg <koen@bergzand.net>
+ * @}
+ */
+
+#include "msg.h"
+#include "tsrb.h"
+
+#include "event.h"
+#include "kernel_defines.h"
+#include "luid.h"
+#include "net/ethernet.h"
+#include "fmt.h"
+#include "usb/descriptor.h"
+#include "usb/cdc.h"
+#include "usb/usbus.h"
+#include "usb/usbus/control.h"
+#include "usb/usbus/cdc/ecm.h"
+
+#include <string.h>
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+static void _event_handler(usbus_t *usbus, usbus_handler_t *handler,
+                          usbus_event_usb_t event);
+static int _setup_handler(usbus_t *usbus, usbus_handler_t *handler,
+                          usbus_setuprq_state_t state, usb_setup_t *setup);
+static void _transfer_handler(usbus_t *usbus, usbus_handler_t *handler,
+                              usbdev_ep_t *ep, usbus_event_transfer_t event);
+static void _init(usbus_t *usbus, usbus_handler_t *handler);
+static void _handle_rx_flush_ev(event_t *ev);
+static void _handle_tx_xmit(event_t *ev);
+
+static size_t _gen_union_descriptor(usbus_t *usbus, void *arg);
+
+static const usbus_hdr_gen_funcs_t _union_descriptor = {
+    .get_header = _gen_union_descriptor,
+    .len = {
+        .fixed_len = sizeof(usb_desc_union_t),
+    },
+    .len_type = USBUS_HDR_LEN_FIXED,
+};
+
+static size_t _gen_ecm_descriptor(usbus_t *usbus, void *arg);
+
+static const usbus_hdr_gen_funcs_t _ecm_descriptor = {
+    .get_header = _gen_ecm_descriptor,
+    .len = {
+        .fixed_len = sizeof(usb_desc_ecm_t),
+    },
+    .len_type = USBUS_HDR_LEN_FIXED,
+};
+
+static size_t _gen_cdc_descriptor(usbus_t *usbus, void *arg);
+
+static const usbus_hdr_gen_funcs_t _cdc_descriptor = {
+    .get_header = _gen_cdc_descriptor,
+    .len = {
+        .fixed_len = sizeof(usb_desc_cdc_t),
+    },
+    .len_type = USBUS_HDR_LEN_FIXED,
+};
+
+static size_t _gen_union_descriptor(usbus_t *usbus, void *arg)
```

Comment:
Isn't the same descriptor for ACM ? maybe we can shared them ?

Reply:
Yup, Either in this PR or as an follow up it can be a simple wrapper to
```C
size_t usbus_cdc_gen_union_descriptor(usbus_t *usbus, usbus_interface_t *master, usbus_interface_t *slave);
```